notebooks/nyc_taxi_years_correlation.ipynb

```diff
@@ -0,0 +1,412 @@
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# tzones = gpd.GeoDataFrame.from_file('tzones_lonlat.json')\n",
+    "# tzones.to_file('cu_taxi_zones.shp')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "taxi_zones = cuspatial.read_polygon_shapefile('cu_taxi_zones.shp')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Converting lon/lat coordinates to LocationIDs with cuSpatial\n",
+    "Looking at the taxi zones and the taxi2016 data, you can see that\n",
+    "- 10.09 million pickup locations\n",
+    "- 10.09 million dropoff locations\n",
+    "- 263 LocationID features\n",
+    "- 354 LocationID rings\n",
+    "- 98,192 LocationID coordinates\n",
+    "\n",
+    "Now that we've collected the set of pickup locations and dropoff locations, we can use `cuspatial.point_in_polygon` to quickly determine which pickups and dropoffs occur in each borough.\n",
+    "\n",
+    "To do this in a memory efficient way, instead of creating two massive 10.09 million x 263 arrays, we're going to use the 31 polygon limit to our advantage and map the resulting true values in the array a new `PULocationID` and `DOLocationID`, matching the 2017 schema.  Locations outside of the `LocationID` areas are `264` and `265`.  We'll be using 264 to indicate our out-of-bounds zones."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[0, 31, 62, 93, 124, 155, 186, 217, 248, 263]\n"
+     ]
+    }
+   ],
+   "source": [
+    "pip_iterations = list(np.arange(0, 263, 31))\n",
+    "pip_iterations.append(263)\n",
+    "print(pip_iterations)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 11.2 s, sys: 6.27 s, total: 17.5 s\n",
+      "Wall time: 17.7 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "taxi2016['PULocationID'] = 264\n",
+    "taxi2016['DOLocationID'] = 264\n",
+    "for i in range(len(pip_iterations)-1):\n",
+    "    start = pip_iterations[i]\n",
+    "    end = pip_iterations[i+1]\n",
+    "    pickups = cuspatial.point_in_polygon(taxi2016['pickup_longitude'] , taxi2016['pickup_latitude'], taxi_zones[0][start:end], taxi_zones[1], taxi_zones[2]['x'], taxi_zones[2]['y'])\n",
+    "    dropoffs = cuspatial.point_in_polygon(taxi2016['dropoff_longitude'] , taxi2016['dropoff_latitude'], taxi_zones[0][start:end], taxi_zones[1], taxi_zones[2]['x'], taxi_zones[2]['y'])\n",
+    "    for j in pickups.columns:\n",
+    "        taxi2016['PULocationID'].loc[pickups[j]] = j\n",
+    "    for j in dropoffs.columns:\n",
+    "        taxi2016['DOLocationID'].loc[dropoffs[j]] = j"
```

Comment:
This takes 2min 28s for me. Is this expected?

Reply:
Yeah, it is expected.  It has to do with the GPU that you are using.  GV100s are the fastest.  My RTX8000 takes 3 minutes.